TheMovieDb/Network/MovieDBNetworkManager.swift

```diff
@@ -0,0 +1,74 @@
+//
+//  NetworkManager.swift
+//  TheMovieDb
+//
+//  Created by Jonathan Hernandez on 03/11/21.
+//
+
+import Foundation
+
+protocol APIClient {
+    
+    var session: URLSession { get }
+    
+    func fetch<T: Decodable>(with request: URLRequest,
+                             decode: @escaping (Decodable) -> T?, completion: @escaping (Result<T, APIError>) -> Void)
+    
+}
+
+extension APIClient {
+    
+    typealias JSONTaskCompletionHandler = (Decodable?, APIError?) -> Void
+    
+    private func decodingTask<T: Decodable>(with request: URLRequest,
+                                            decodingType: T.Type,
+                                            completionHandler completion: JSONTaskCompletionHandler?) -> URLSessionDataTask {
+        let task = session.dataTask(with: request) { data, response, _ in
+            guard let httpResponse = response as? HTTPURLResponse else {
+                completion?(nil, .requestFailed)
+                return
+            }
+            if httpResponse.statusCode == 200 || httpResponse.statusCode == 201 {
+                if let data = data {
+                    do {
+                        let decoder = JSONDecoder()
+                        let genericModel = try decoder.decode(decodingType, from: data)
+                        completion?(genericModel, nil)
+                    } catch {
+                        print(error.localizedDescription)
+                        completion?(nil, .requestFailed)
+                    }
+                } else {
+                    completion?(nil, .invalidData)
+                }
+            } else {
+                completion?(nil, APIError(response: httpResponse))
+            }
+        }
+        return task
+    }
+    
+    func fetch<T: Decodable>(with request: URLRequest,
+                             decode: @escaping (Decodable) -> T?,
+                             completion: @escaping (Result<T, APIError>) -> Void) {
+        let task = decodingTask(with: request, decodingType: T.self) { (json, error) in
+            DispatchQueue.main.async {
```

Comment:
Just a question: Should the APIClient be responsible for handling the response in the main thread?